plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/GcsTransactionLogSynchronizer.java

```diff
@@ -0,0 +1,126 @@
+import com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.CreateObjectOptions;
+import com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageItemInfo;
+import com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.StorageResourceId;
+import com.google.common.collect.Maps;
+import io.trino.plugin.deltalake.GcsStorageFactory;
+import io.trino.spi.connector.ConnectorSession;
+import org.apache.hadoop.fs.Path;
+
+import javax.inject.Inject;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.util.Base64;
+import java.util.Map;
+
+import static com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageFileSystem.objectOptionsFromFileOptions;
+import static java.util.Objects.requireNonNull;
+
+public class GcsTransactionLogSynchronizer
+        implements TransactionLogSynchronizer
+{
+    private final GcsStorageFactory gcsStorageFactory;
+
+    @Inject
+    public GcsTransactionLogSynchronizer(GcsStorageFactory gcsStorageFactory)
+    {
+        this.gcsStorageFactory = requireNonNull(gcsStorageFactory, "gcsStorageFactory is null");
+    }
+
+    // This approach should be compatible with OSS Delta Lake
+    // https://github.com/delta-io/delta/blob/master/storage/src/main/java/io/delta/storage/GCSLogStore.java
+    // This approach differs from the OSS Delta Lake corresponding implementation
+    // in the sense that it uses the `Storage` GCS API client directly (instead of Hadoop's `FSDataOutputStream`)
+    // in order to avoid leaked output streams in case of I/O exceptions occurring while uploading
+    // the blob content.
+    @Override
+    public void write(ConnectorSession session, String clusterId, Path newLogEntryPath, byte[] entryContents)
+    {
+        Storage storage = gcsStorageFactory.create(session);
+        try {
+            createStorageObjectExclusively(newLogEntryPath.toUri(), entryContents, storage);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public boolean isUnsafe()
+    {
+        return false;
+    }
+
+    /**
+     * This method uses {@link Storage} Cloud Storage API client for creating synchronously a storage object
+     * with the specified path and content.
+     * <p>
+     * This approach avoids dealing with `FSDataOutputStream` exposed by the method
+     * `org.apache.hadoop.fs.FileSystem#create(org.apache.hadoop.fs.Path)` and the intricacies of handling
+     * exceptions which may occur while writing the content to the output stream.
+     */
+    private static StorageObject createStorageObjectExclusively(URI blobPath, byte[] content, Storage storage)
+            throws IOException
+    {
+        CreateFileOptions createFileOptions = CreateFileOptions.DEFAULT_NO_OVERWRITE;
```

Comment:
Distinction between `CreateFileOptions.DEFAULT_NO_OVERWRITE` and `CreateFileOptions#DEFAULT_OVERWRITE` is in their `CreateFileOptions#isOverwriteExisting`.
Where is it used?

I understand that the no-overwrite behavior is effected by `insert.setIfGenerationMatch(0L);` line, so does this `CreateFileOptions` actually matter?

Reply:
Good point. 

I actually copied and adapted initially the code from gcs hadoop-connectors project.
I notice now that much of this code is not actually needed.

I removed the creation of the `CreateFileOptions` instance altogether.